lib/spack/spack/binary_distribution.py

```diff
@@ -935,15 +935,22 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
     spackfile_path = os.path.join(
         cache_prefix, tarball_path_name(spec, '.spack'))
 
-    remote_spackfile_path = url_util.join(
+    outdir = mirror.push_url
+    fetch_spackfile_path = url_util.join(
+        mirror.fetch_url, os.path.relpath(spackfile_path, tmpdir))
+    push_spackfile_path = url_util.join(
         outdir, os.path.relpath(spackfile_path, tmpdir))
 
     mkdirp(tarfile_dir)
-    if web_util.url_exists(remote_spackfile_path):
+    if web_util.url_exists(push_spackfile_path):
         if force:
-            web_util.remove_url(remote_spackfile_path)
+            web_util.remove_url(push_spackfile_path)
         else:
-            raise NoOverwriteException(url_util.format(remote_spackfile_path))
+            raise NoOverwriteException(url_util.format(push_spackfile_path))
+
+    if web_util.url_exists(fetch_spackfile_path):
```

Comment:
I was thinking if you did like I suggested above, you can then remove this second block checking the url existence a second time.